pkg/services/cosmosdb/cosmos-utils.go

```diff
@@ -178,3 +182,216 @@ func deleteDatabase(
 	}
 	return nil
 }
+
+// The deployment will return success once the write region is created,
+// ignoring the status of read regions , so we must implement detection logic
+// by ourselves.
+func (c *cosmosAccountManager) waitForReadLocationsReady(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*cosmosdbInstanceDetails)
+	resourceGroupName := instance.ProvisioningParameters.GetString("resourceGroup")
+	accountName := dt.DatabaseAccountName
+	databaseAccountClient := c.databaseAccountsClient
+
+	err := pollingUntilReadLocationsReady(
+		ctx,
+		resourceGroupName,
+		accountName,
+		databaseAccountClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+// For sqlAllInOneManager, the real type of `instance.Details` is
+// `*sqlAllInOneInstanceDetails`, so type assertion must be changed.
+// Expect type assertion, this function is totally the same as previous one.
+func (s *sqlAllInOneManager) waitForReadLocationsReady(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*sqlAllInOneInstanceDetails)
+	resourceGroupName := instance.ProvisioningParameters.GetString("resourceGroup")
+	accountName := dt.DatabaseAccountName
+	databaseAccountClient := s.databaseAccountsClient
+
+	err := pollingUntilReadLocationsReady(
+		ctx,
+		resourceGroupName,
+		accountName,
+		databaseAccountClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+const succeeded = "succeeded"
+
+// For now, this method will return on either context is cancelled or
+// every region's state is "succeeded" in seven consecutive check.
+// The reason why we need seven consecutive check is that
+// the read region is created one by one, there is a small gap between
+// the finishment of previous creation and the start of the next creation.
+// By this check, we can detect gaps shorter than 1 mintue,
+// and report success within 70 seconds after completion.
```

Comment:
This heuristic will not work.

It seems implied here that read locations are added and become ready one-by-one (i.e. in serial).

1. Is that even true? That seems _really_ odd. My gut would expect that `result.DatabaseAccountProperties.ReadLocations` consistently reports the same number of read locations, regardless of their state. Shouldn't we just expect that if n locations were requested that each time we poll, some m =< n will be ready and some n - m will _not_ be ready? If that's the case, shouldn't we only have to poll until all n are reported as ready _just once_ ? (Maybe this isn't how this API works?? Correct me if I am wrong.)

2. Let's assume I am totally wrong about no.1 above. So let's say that `result.DatabaseAccountProperties.ReadLocations` does _not_ consistently report the same number of read locations, regardless of their state. In such a case, don't we only need to wait until read regions m == desired read regions n _and_ all n are ready?

Which of these is correct depends a lot on the mechanics of the underlying API, and I'm not an expert on that... but whichever is the case, one of the two above will prove to be a logically correct algorithm that is better than the heuristic currently used here... and either will return faster in the successful case than this heuristic would.

Additionally-- some concession probably needs to be made for a timeout here. There should be a finite amount of time we will wait for all read locations to be ready before giving up.

Reply:
Hi @krancour , `result.DatabaseAccountProperties.ReadLocations` works in following way: 
Say I provision an instance with one write location `A` and three read locations `B,C,D`. At first, when `A` is creating, the function will return only one read location, `A` with state "creating'; when the creation of `A` is finished, creation of `B` will start, now the function will return two read locations, `A` with state "succeeded" and `B` with state "creating"; when the creation of `B` is finished, creation of `C` will start, now it will returns three locations, `A` and `B` with state "succeeded" and `C` with state `creating`, and so on.
You are right, it's better to check the the number of read locations and their state. I'm trapped in conventional thinking and it's actually not a good design.

As for a timeout, it has a very wide range based on how many read regions the user creates. How about estimate an upper bound based on the number of read regions? Say the user creates 3 regions, then we set the time out to `3 * 15 * time.Minutes`?

Reply:
That seems like a huge amount of time to wait for something like this to complete. Does it really take that long? Is there no way to bring up the read regions in parallel?

Reply:
Hi @krancour , for now, we can't make the creation in parallel. The underlying creation is managed by Cosmos team, currently the creation is serial on both Azure portal and go-sdk.

Reply:
And 15 mintues for one read location is a pessimistic estimation, I will investigate it and find a more precise bound.